### ISSUE TITLE
nonlinear_bellman.py

### ISSUE THREAD

**Issue by YHL04** [michaelnny/deep_rl_zoo#15]
c_t = torch.minimum(torch.tensor(1.0), pi_a_t / (mu_t + eps)) * lambda_

mu_t should be mu_a_t right? Is this a bug or intentional?

**Comment by YHL04** [comment]
discount_t = (~done).float() * discount  # (T+1, B)

        # Derive target policy probabilities from q values.
        target_policy_probs = F.softmax(q_t, dim=-1)  # [T+1, B, action_dim]

        if self._transformed_retrace:
            transform_tx_pair = nonlinear_bellman.SIGNED_HYPERBOLIC_PAIR
        else:
            transform_tx_pair = nonlinear_bellman.IDENTITY_PAIR  # No transform

        # Compute retrace loss.
        retrace_out = nonlinear_bellman.transformed_retrace(
            q_tm1=q_t[:-1],
            q_t=target_q_t[1:],
            a_tm1=a_t[:-1],
            a_t=a_t[1:],
            r_t=r_t[:-1],
            discount_t=discount_t[:-1],
            pi_t=target_policy_probs[1:],
            mu_t=behavior_prob_a_t[1:],
            lambda_=self._retrace_lambda,
            tx_pair=transform_tx_pair,
        )

since done will only be true at the end of the episode,

discount_t = (~done).float() * discount  # (T+1, B)

done will only be true at the end of T+1
so discount[:-1] will only contain discount, so the line above is pointless
will that be a big issue in training if agent never trained on the terminal timestep
where discount = 0?

**Comment by YHL04** [comment]
@michaelnny 

target_policy_probs = F.softmax(q_t, dim=-1)  # [T+1, B, action_dim]

should be 

target_policy_probs = F.softmax(target_q_t, dim=-1)  # [T+1, B, action_dim]

from Agent57 paper pg17: "π(a|x) is the target policy"

**Comment by michaelnny** [comment]
Hi, I will take a look at it once I have some free time.
